// @microsoft/teams.apps/dist/middleware/with-client-auth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withClientAuth = withClientAuth;
function withClientAuth(params) {
    const entraTokenValidator = params.entraTokenValidator;
    const log = params.logger;
    return async (req, res, next) => {
        const appSessionId = req.header('X-Teams-App-Session-Id');
        const pageId = req.header('X-Teams-Page-Id');
        const authorization = req.header('Authorization')?.split(' ');
        const authToken = authorization?.length === 2 && authorization[0].toLowerCase() === 'bearer'
            ? authorization[1]
            : '';
        const validatedToken = !entraTokenValidator
            ? null
            : await entraTokenValidator.validateAccessToken(log, authToken);
        const tokenPayload = validatedToken && entraTokenValidator?.getTokenPayload(validatedToken);
        if (!pageId ||
            !appSessionId ||
            !validatedToken ||
            !entraTokenValidator ||
            !tokenPayload) {
            log.debug('unauthorized');
            res.status(401).send('unauthorized');
            return;
        }
        req.context = {
            appId: tokenPayload?.['appId'],
            appSessionId,
            authToken,
            channelId: req.header('X-Teams-Channel-Id'),
            chatId: req.header('X-Teams-Chat-Id'),
            meetingId: req.header('X-Teams-Meeting-Id'),
            messageId: req.header('X-Teams-Message-Id'),
            pageId,
            subPageId: req.header('X-Teams-Sub-Page-Id'),
            teamId: req.header('X-Teams-Team-Id'),
            tenantId: tokenPayload['tid'],
            userId: tokenPayload['oid'],
            userName: tokenPayload['name'],
        };
        next();
    };
}
